File: frontend/src/pages/StoreHealth.js
import { useEffect, useState } from "react";
import { toast } from "sonner";
import api from "@/lib/api";
import { PageHeader } from "@/components/shared";
import {
  Loader2, RefreshCw, CheckCircle2, AlertTriangle, XCircle, Store,
  BookOpen, ShieldCheck, Mail, ShoppingCart, Sparkles,
} from "lucide-react";

const STATUS = {
  healthy: { cls: "bg-emerald-100 text-emerald-700 border-emerald-200", Icon: CheckCircle2, label: "Healthy" },
  warnings: { cls: "bg-amber-100 text-amber-700 border-amber-200", Icon: AlertTriangle, label: "Warnings" },
  attention: { cls: "bg-red-100 text-red-700 border-red-200", Icon: XCircle, label: "Needs attention" },
};

function Stat({ label, value, tone = "default" }) {
  const tones = { default: "bg-card", ok: "bg-emerald-50 text-emerald-700", warn: "bg-amber-50 text-amber-700", bad: "bg-red-50 text-red-700" };
  return (
    <div className={`rounded-lg border p-4 ${tones[tone]}`}>
      <div className="text-2xl font-bold">{value}</div>
      <div className="text-[11px] mt-1 opacity-80">{label}</div>
    </div>
  );
}

function Section({ icon: Icon, title, children }) {
  return (
    <section className="rounded-xl border bg-card p-5">
      <div className="flex items-center gap-2 mb-4">
        <Icon className="w-4 h-4 text-navy" />
        <span className="font-heading font-bold text-navy">{title}</span>
      </div>
      {children}
    </section>
  );
}

export default function StoreHealth() {
  const [data, setData] = useState(null);
  const [busy, setBusy] = useState(false);

  const load = () => {
    setBusy(true);
    api.get("/store/health")
      .then((r) => setData(r.data))
      .catch((e) => { setData(false); toast.error(e.response?.data?.detail || "Failed to load store health."); })
      .finally(() => setBusy(false));
  };
  useEffect(() => { load(); }, []);

  if (data === null) return <div className="flex justify-center py-32"><Loader2 className="w-6 h-6 animate-spin text-royal" /></div>;
  if (data === false) return <div className="text-sm text-muted-foreground py-16 text-center">Unable to load store health.</div>;

  const o = data.overall || {};
  const sf = data.storefront || {};
  const hy = data.hygiene || {};
  const cm = data.commerce || {};
  const or = data.orders || {};
  const st = STATUS[o.status] || STATUS.warnings;

  return (
    <div className="space-y-6" data-testid="store-health">
      <div className="flex items-start justify-between gap-3 flex-wrap">
        <PageHeader title="Store Health™" subtitle="A live, evidence-based view of your storefront's integrity. Every number is read straight from this environment's database — no fabricated metrics." />
        <button onClick={load} disabled={busy} data-testid="refresh-health" className="inline-flex items-center gap-2 rounded-lg border border-navy/30 text-navy px-4 py-2 text-sm font-medium hover:bg-navy/5 disabled:opacity-50">
          {busy ? <Loader2 className="w-4 h-4 animate-spin" /> : <RefreshCw className="w-4 h-4" />} Refresh
        </button>
      </div>

      {/* Overall banner */}
      <div className={`rounded-xl border-2 p-5 flex items-center gap-4 ${st.cls}`} data-testid="overall-status">
        <st.Icon className="w-8 h-8 shrink-0" />
        <div className="flex-1">
          <div className="font-heading font-bold text-lg">{st.label}</div>
          <div className="text-sm opacity-90">{o.passed}/{o.total} integrity checks passing</div>
        </div>
        <Store className="w-10 h-10 opacity-30 shrink-0" />
      </div>

      {/* Checks */}
      <Section icon={ShieldCheck} title="Integrity Checks">
        <div className="space-y-2" data-testid="health-checks">
          {(o.checks || []).map((c) => (
            <div key={c.key} className="flex items-center gap-3 rounded-lg border p-3" data-testid={`check-${c.key}`}>
              {c.ok ? <CheckCircle2 className="w-4 h-4 text-emerald-600 shrink-0" /> : <AlertTriangle className="w-4 h-4 text-amber-600 shrink-0" />}
              <div className="flex-1 text-sm text-foreground">{c.label}</div>
              <span className={`text-[11px] font-semibold rounded-full px-2.5 py-1 ${c.ok ? "bg-emerald-100 text-emerald-700" : "bg-amber-100 text-amber-700"}`}>{c.detail}</span>
            </div>
          ))}
        </div>
      </Section>

      {/* Storefront */}
      <Section icon={BookOpen} title="Public Storefront">
        <div className="grid grid-cols-2 lg:grid-cols-4 gap-3">
          <Stat label="Authorized books" value={sf.authorized_books} />
          <Stat label="Live on storefront" value={sf.live_on_storefront} tone="ok" />
          <Stat label="Hidden (missing cover)" value={sf.hidden_missing_cover} tone={sf.hidden_missing_cover ? "bad" : "ok"} />
          <Stat label={`Featured (${sf.featured_mode})`} value={sf.featured_selected} />
        </div>
        {sf.hidden_missing_cover > 0 && (
          <div className="mt-3 text-[11px] text-red-700">Hidden titles (no cover → not shown publicly): {(sf.missing_cover_titles || []).join(", ")}</div>
        )}
      </Section>

      {/* Hygiene */}
      <Section icon={Sparkles} title="Catalog Hygiene">
        <div className="grid grid-cols-2 lg:grid-cols-3 gap-3">
          <Stat label="Test products detected" value={hy.test_products_matched} />
          <Stat label="Test products visible in catalog" value={hy.test_products_visible_in_catalog} tone={hy.test_products_visible_in_catalog ? "warn" : "ok"} />
        </div>
        {hy.test_products_visible_in_catalog > 0 && (
          <div className="mt-3 text-[11px] text-amber-700">
            {hy.test_products_visible_in_catalog} internal test products are visible to learners. Run <span className="font-medium">Test Product Cleanup</span> in Production Operations™ to archive them.
          </div>
        )}
      </Section>

      {/* Commerce + Orders */}
      <div className="grid lg:grid-cols-2 gap-6">
        <Section icon={ShoppingCart} title="Commerce Configuration">
          <div className="space-y-2 text-sm">
            <Row label="Stripe mode" value={cm.stripe_mode} tone={cm.stripe_mode === "LIVE" ? "live" : cm.stripe_mode === "TEST" ? "test" : "bad"} />
            <Row label="Accepts real money" value={cm.accepts_real_money ? "Yes" : "No"} />
            <Row label="Webhook secret" value={cm.webhook_secret_set ? "Set" : "Missing"} tone={cm.webhook_secret_set ? "ok" : "warn"} />
            <Row label="Email provider connected" value={cm.email_provider_key_set ? "Yes" : "No"} tone={cm.email_provider_key_set ? "ok" : "warn"} />
            <Row label="Sender email" value={cm.sender_email_set ? "Set" : "Missing"} tone={cm.sender_email_set ? "ok" : "warn"} />
          </div>
        </Section>
        <Section icon={Mail} title="Orders & Delivery">
          <div className="grid grid-cols-2 gap-3">
            <Stat label="Orders total" value={or.orders_total} />
            <Stat label="Paid" value={or.orders_paid} tone="ok" />
            <Stat label="Pending" value={or.orders_pending} />
            <Stat label="Confirmations failed" value={or.confirmation_failed} tone={or.confirmation_failed ? "bad" : "ok"} />
            <Stat label="Confirmations sent" value={or.confirmation_sent} />
            <Stat label="Newsletter subscribers" value={or.newsletter_subscribers} />
          </div>
        </Section>
      </div>

      <p className="text-[11px] text-muted-foreground">Snapshot at {new Date(data.at).toLocaleString()} · reflects the database this environment is connected to (production when viewed on the live site).</p>
    </div>
  );
}

function Row({ label, value, tone }) {
  const tones = { ok: "text-emerald-700", warn: "text-amber-700", bad: "text-red-700", live: "text-emerald-700", test: "text-blue-700" };
  return (
    <div className="flex items-center justify-between rounded-lg border px-3 py-2">
      <span className="text-muted-foreground">{label}</span>
      <span className={`font-semibold ${tones[tone] || "text-foreground"}`}>{value}</span>
    </div>
  );
}

File: backend/store_health.py
"""QRU Store Health™ — read-only, evidence-based integrity view of the live store.

Treasure Standard™: every number is derived from real data in THIS environment's
database (production when run on the deployed site). No fabricated metrics, no
fake "healthy" states. Purely read-only — computes, never writes.
"""
import os
from datetime import datetime, timezone

from database import db
import prod_migrations as pm

_PUBLISHED_QUERY = {"founder_authorization.authorized": True}


def _now():
    return datetime.now(timezone.utc).isoformat()


def _cover_url(book: dict):
    design = (book.get("artifacts") or {}).get("design") or {}
    concepts = design.get("cover_concepts") or []
    selected = design.get("selected_cover") or {}
    sel_no = selected.get("concept")
    if sel_no is not None:
        for c in concepts:
            if c.get("concept") == sel_no and c.get("url"):
                return c["url"]
    for c in concepts:
        if c.get("url"):
            return c["url"]
    return None


async def _storefront():
    docs = await db.book_records.find(_PUBLISHED_QUERY, {"_id": 0}).to_list(1000)
    with_cover, missing = [], []
    for b in docs:
        (with_cover if _cover_url(b) else missing).append(b)
    settings = await db.storefront_settings.find_one({"key": "featured"})
    featured = (settings or {}).get("book_ids") or []
    return {
        "authorized_books": len(docs),
        "live_on_storefront": len(with_cover),          # only cover-bearing books show publicly
        "hidden_missing_cover": len(missing),
        "missing_cover_titles": [b.get("title") for b in missing][:20],
        "featured_selected": len(featured),
        "featured_mode": "founder-curated" if featured else "auto",
    }


async def _hygiene():
    rows = await pm._test_products_classify()
    visible = [r for r in rows if r["classification"] == "TEST_VISIBLE_IN_CATALOG"]
    return {
        "test_products_matched": len(rows),
        "test_products_visible_in_catalog": len(visible),
        "visible_codes": [r["code"] for r in visible][:20],
    }


def _commerce_config():
    key = os.environ.get("STRIPE_API_KEY", "")
    test_override = os.environ.get("STRIPE_TEST")
    if key.startswith("sk_live"):
        mode = "LIVE"
    elif key.startswith("sk_test") or key == "sk_test_emergent" or test_override:
        mode = "TEST"
    else:
        mode = "UNKNOWN"
    return {
        "stripe_mode": mode,
        "accepts_real_money": mode == "LIVE",
        "webhook_secret_set": bool(os.environ.get("STRIPE_WEBHOOK_SECRET")),
        "sender_email_set": bool(os.environ.get("SENDER_EMAIL")),
        "support_email_set": bool(os.environ.get("SUPPORT_EMAIL")),
        "email_provider_key_set": bool(os.environ.get("RESEND_API_KEY")),
    }


async def _orders():
    total = await db.book_purchases.count_documents({})
    paid = await db.book_purchases.count_documents({"payment_status": "paid"})
    pending = await db.book_purchases.count_documents({"payment_status": "pending"})
    # email delivery breakdown for paid orders
    sent = await db.book_purchases.count_documents(
        {"payment_status": "paid", "confirmation_email.status": {"$in": ["sent-to-provider", "sent", "delivered"]}})
    skipped = await db.book_purchases.count_documents(
        {"payment_status": "paid", "confirmation_email.status": "skipped"})
    failed = await db.book_purchases.count_documents(
        {"payment_status": "paid", "confirmation_email.status": "error"})
    subs = await db.newsletter_subscribers.count_documents({})
    return {
        "orders_total": total, "orders_paid": paid, "orders_pending": pending,
        "confirmation_sent": sent, "confirmation_skipped": skipped, "confirmation_failed": failed,
        "newsletter_subscribers": subs,
    }


async def _cover_integrity():
    """Live durable check of every published book's ACTIVE cover master. Flags books whose
    active cover is missing from durable storage (the root cause of broken cover thumbnails)
    and whether a Founder-generated alternate concept is available to fall back to."""
    docs = await db.book_records.find(_PUBLISHED_QUERY, {"_id": 0}).to_list(1000)
    broken, repairable, unrepairable, checked = [], 0, 0, 0
    for b in docs:
        design = (b.get("artifacts") or {}).get("design") or {}
        concepts = design.get("cover_concepts") or []
        selected = design.get("selected_cover") or {}
        url = selected.get("url") or _cover_url(b)
        if not url:
            continue  # missing-cover books are reported under storefront
        checked += 1
        fname = url.rsplit("/", 1)[-1]
        if await pm._asset_ok(fname):
            continue
        alt = None
        for c in concepts:
            cu = c.get("url")
            if not cu or cu.rsplit("/", 1)[-1] == fname or c.get("status") == "failed":
                continue
            if await pm._asset_ok(cu.rsplit("/", 1)[-1]):
                alt = c.get("concept")
                break
        rep = alt is not None
        repairable += 1 if rep else 0
        unrepairable += 0 if rep else 1
        broken.append({"code": b.get("book_code"), "title": b.get("title"),
                       "active_cover": fname, "repairable": rep, "fallback_concept": alt})
    return {"checked": checked, "broken": len(broken), "repairable": repairable,
            "unrepairable": unrepairable, "books": broken[:50]}


def _overall(storefront, hygiene, commerce, orders, cover=None):
    """Deterministic health rollup — honest signals, not a vanity score."""
    checks = []
    checks.append({"key": "storefront_has_books", "ok": storefront["live_on_storefront"] > 0,
                   "label": "At least one book is live on the storefront",
                   "detail": f"{storefront['live_on_storefront']} live"})
    checks.append({"key": "no_hidden_covers", "ok": storefront["hidden_missing_cover"] == 0,
                   "label": "No authorized book is hidden by a missing cover",
                   "detail": f"{storefront['hidden_missing_cover']} hidden"})
    checks.append({"key": "no_visible_test_products", "ok": hygiene["test_products_visible_in_catalog"] == 0,
                   "label": "No internal test products visible in the learner catalog",
                   "detail": f"{hygiene['test_products_visible_in_catalog']} visible"})
    checks.append({"key": "checkout_configured", "ok": commerce["stripe_mode"] != "UNKNOWN",
                   "label": "Stripe checkout is configured",
                   "detail": f"{commerce['stripe_mode']} mode"})
    checks.append({"key": "webhook_secret", "ok": commerce["webhook_secret_set"],
                   "label": "Stripe webhook secret is set (secure fulfillment)",
                   "detail": "set" if commerce["webhook_secret_set"] else "missing"})
    checks.append({"key": "email_delivery", "ok": commerce["email_provider_key_set"],
                   "label": "Confirmation email provider is connected",
                   "detail": "connected" if commerce["email_provider_key_set"] else "not connected"})
    checks.append({"key": "no_failed_confirmations", "ok": orders["confirmation_failed"] == 0,
                   "label": "No failed purchase-confirmation emails",
                   "detail": f"{orders['confirmation_failed']} failed"})
    if cover is not None:
        checks.append({"key": "covers_durable", "ok": cover["broken"] == 0,
                       "label": "Every live book's active cover is retrievable from durable storage",
                       "detail": "all durable" if cover["broken"] == 0 else f"{cover['broken']} broken ({cover['repairable']} repairable)"})
    passed = sum(1 for c in checks if c["ok"])
    critical = {"storefront_has_books", "no_visible_test_products", "checkout_configured"}
    critical_fail = [c["key"] for c in checks if not c["ok"] and c["key"] in critical]
    if critical_fail:
        status = "attention"
    elif passed == len(checks):
        status = "healthy"
    else:
        status = "warnings"
    return {"status": status, "passed": passed, "total": len(checks), "checks": checks}


async def report():
    storefront = await _storefront()
    hygiene = await _hygiene()
    commerce = _commerce_config()
    orders = await _orders()
    cover = await _cover_integrity()
    overall = _overall(storefront, hygiene, commerce, orders, cover)
    return {
        "at": _now(),
        "overall": overall,
        "storefront": storefront,
        "hygiene": hygiene,
        "commerce": commerce,
        "orders": orders,
        "cover_integrity": cover,
    }
